### PR TITLE
Memoizing DataGens in integration tests

### LINKED ISSUE #6201: [FEA] experiment with memoizing datagens in the integration_tests
Given a `DataGen` "schema" it should be possible to save some (a lot?) of time in the integration tests if we don't regenerate the data twice (once for cpu and for gpu) x all the tests we have. This is something that came up when discussing with @revans2 where he has noticed quite a bit of time spent on the data generation in the tests.

This task would be to figure out how to memoize (we'd need to take into account the seed, the columns being generated, and any nestedness), and see if we can cut down the time of the tests.

### PR DESCRIPTION
**Pull request by thirtiseven**
Closes #6201

This PR cached `gen_df` function in integration tests to speed up dataframes generating, which was reported costing quite a bit of time. ~~The mainly change is implement a `hash_object` function to hash `DataGen`s so they can be cached correctly.~~ The mainly change is adding `_cache_repr` in `DataGen`s and comparing them in `__eq__` so they can be cached correctly.

#### Tests

Note: the results are outdated (again) since I added sparkconf as a cache key, will check if this is necessary and update the test results. An IT local results of this change: 1:13:22 CacheInfo(hits=1633, misses=4371, maxsize=128, currsize=128)

```
TEST_PARALLEL=1 ./run_pyspark_from_build.sh -k 'cmp'
```
before:
430 passed, 19544 deselected, 1 warning in 312.23s (0:05:12)
after:
430 passed, 19545 deselected, 1 warning in 188.93s (0:03:08)

Local Integration test with default configs:
before:
18234 passed, 848 skipped, 630 xfailed, 262 xpassed, 96 warnings in 4772.34s (1:19:32)
after:
18235 passed, 848 skipped, 630 xfailed, 262 xpassed, 97 warnings in 3584.41s (0:59:44)

CI Took 1 hr 55 min / 2 hr 10 min (other recent CI took about 2 hr 35 min).

Some cache results:
None: CacheInfo(hits=3181, misses=664, maxsize=None, currsize=600)
128:  CacheInfo(hits=3873, misses=1550, maxsize=128, currsize=128)

The time taken for different maxsizes is similar, so I set it to 128 for memory saving.
Note that the (hits+misses) total count may be different because of parallel testing.

About additional memory usage. I tested that the max memory usage of a single test of all current IT is 7282688 bytes. so the LRU cache of size 128 may cost additional 7282688 * 128 bytes = 932 MB memory if every slots with huge data sets.

### CODE REVIEW

**Comment by thirtiseven** [comment]
build

**Comment by thirtiseven** [comment]
build

**Comment by thirtiseven** [comment]
build

**Review by revans2 — DISMISSED** [review]
This looks great!!! I ran

```
TEST_PARALLEL=1 ./run_pyspark_from_build.sh -k 'cmp'
```
and it sped things up a lot.

before!!!
========= 430 passed, 18599 deselected, 1 warning in 344.83s (0:05:44) =========
after!!!!
========= 430 passed, 18599 deselected, 1 warning in 223.66s (0:03:43) =========

That is a 35% performance improvement!!!!

File: integration_tests/src/main/python/data_gen.py [hunk 1]
```diff
@@ -27,12 +27,43 @@
...
 # set time zone to UTC for timestamp test cases to avoid `datetime` out-of-range error:
 # refer to: https://github.com/NVIDIA/spark-rapids/issues/7535
 os.environ['TZ'] = 'UTC'
 time.tzset()
 
+# hash object with nested data types for caching
+def hash_object(obj):
+    h = hashlib.blake2b(digest_size=10)
+    if isinstance(obj, (int, float, str, bool, type(None))):
+        h.update(str(obj).encode())
```
Comment (abellina) [inline integration_tests/src/main/python/data_gen.py:42]: It might be nice to add a comment on why `str(obj)` is needed. 

I half expected `hash_object` to be a call to an API provided by python or another package and I am curious if `pickle.dumps` accomplishes part of it (not for dictionaries, we'd have to make sure to convert to an ordered dictionary).

**Review by abellina — DISMISSED** [review]
Amazing results. I had a comment that is non blocker

File: integration_tests/src/main/python/data_gen.py (hunk 1 above)
Comment (thirtiseven) [reply integration_tests/src/main/python/data_gen.py:43]: Thanks for review! `update` needs bytes-like object so I convert them to string, then encode to utf-8. Added a comment there.
I think `pickle.dumps` can serialize the DataGen so we can hash on it, but it may need to override the `__reduce__` method to handle things like internal states, lambdas, and datagens nested in dictionaries, so it won't make the solution easier. I guess we need to implement the `hash_object` because it's basically an override to `__hash__` for our custom mutable objects.

**Comment by thirtiseven** [comment]
build

**Comment by thirtiseven** [comment]
build

File: integration_tests/src/main/python/data_gen.py (hunk 1 above)
Comment (abellina) [reply integration_tests/src/main/python/data_gen.py:43]: Also, for `str(obj)` I think part of it is that if these basic types fellthrough to the else:
```
h.update(str(hash(obj)).encode())
```
We could end up with undesired results. It looks like `hash(True)==hash(1)` for example, but `hash(str(True))!=hash(str(1))`. But that said, I think in this case, it might be unlikely we'll find a collision since we are traversing a set number of arguments and we follow a pattern.

**Review by abellina — APPROVED** [review]

File: integration_tests/src/main/python/data_gen.py [hunk 2]
```diff
@@ -27,12 +27,44 @@
...
+        for k in keys:
+            if str(k) == '_gen_func':
+                # bypass internal functions
+                continue
+            h.update(hash_object(k).encode())
+            h.update(hash_object(attrs[k]).encode())
+    elif callable(obj):
+        h.update(obj.__code__.co_code)
+    else:
+        h.update(str(hash(obj)).encode())
```
Comment (gerashegalov) [inline integration_tests/src/main/python/data_gen.py:65]: is this where `spark` SparkSession  is handled?

looks like hash of SparkSession does not reflect the current runtime config. If we can reason it's ok in use case please add some comment. Otherwise, we might need an issue to handle configs affecting createDataFrame

File: integration_tests/src/main/python/data_gen.py (hunk 1 above)
Comment (abellina) [reply integration_tests/src/main/python/data_gen.py:43]: Yes, but in terms of hashing here for caching purposes, you may want to encode the type somehow...

Comment (revans2) [reply integration_tests/src/main/python/data_gen.py:43]: Should we just force each DataGen instance to have a method that returns something that can be a part of the hash, and then it is up to the DataGen to do this properly?  We are trying to make something that is generic for everything, but I don't think that is going to work or else the hashing library would have done it already.

As a side note can use look at simple equality instead? The speedup is mostly going to come from run on CPU then run the same thing on the GPU.  We might not need to solve this generically for everything if we can just check that the two instances are the same, no hashing needed.

Comment (gerashegalov) [reply integration_tests/src/main/python/data_gen.py:43]: > Yes, but in terms of hashing here for caching purposes, you may want to encode the type somehow...

lru_cache has a parameter `typed` which is `False` by default for some reason. But we can activate it. https://github.com/python/cpython/blob/60f8117d0c685c2923b7cb17b725b67cd41e8410/Lib/functools.py#L485

File: integration_tests/src/main/python/data_gen.py [hunk 3]
```diff
@@ -681,6 +713,17 @@ def skip_if_not_utc():
     if (not is_tz_utc()):
         skip_unless_precommit_tests('The java system time zone is not set to UTC')
 
+@lru_cache(maxsize=128)
```
Comment (gerashegalov) [inline integration_tests/src/main/python/data_gen.py:716]: consider
```suggestion
@lru_cache(maxsize=128, typed=True)
```

**Comment by gerashegalov** [comment]
> - maxsize = None: CacheInfo(hits=2820, misses=1023, maxsize=None, currsize=737)
> - maxsize = 512: CacheInfo(hits=2788, misses=968, maxsize=512, currsize=512)
> - maxsize = 256: CacheInfo(hits=3392, misses=1032, maxsize=256, currsize=256)
> - maxsize = 128: CacheInfo(hits=4781, misses=945, maxsize=128, currsize=128)

Why is the number of hits decreasing when max_size is increasing?

File: integration_tests/src/main/python/data_gen.py (hunk 3 above)
Comment (thirtiseven) [reply integration_tests/src/main/python/data_gen.py:716]: Thanks for the review! I don't think the `typed` here can affect the behaviour of `hash_object`, because `@lru_cache` checks if an object is in the cache with the `__eq__` method, and I just use the hash value to check for equality.

File: integration_tests/src/main/python/data_gen.py (hunk 2 above)
Comment (thirtiseven) [reply integration_tests/src/main/python/data_gen.py:65]: `@lru_cache` will handle the lookup of SparkSession in hash with it's `__eq__` method, and the current runtime config will not affect the equality of it. The `hash_object` will only used to check the equality of `DataGen`s.

Comment (gerashegalov) [reply integration_tests/src/main/python/data_gen.py]: > the current runtime config will not affect the equality of it

and thus we need to be able to reason why or when is our use of createDataFrame produces identical results irrespective of the runtime config. There is at least one conditional already at the Python level https://github.com/apache/spark/blob/c3b62708cd6371da08943e572a9bc0a45c1dcced/python/pyspark/sql/session.py#L1332, and maybe more at the underlying Scala as well

File: integration_tests/src/main/python/data_gen.py (hunk 3 above)
Comment (gerashegalov) [reply integration_tests/src/main/python/data_gen.py:733]: having type as an input to cache's hash key makes unexpected side effects from another [thread](https://github.com/NVIDIA/spark-rapids/pull/8441#discussion_r1213172989) go away.

File: integration_tests/src/main/python/data_gen.py (hunk 1 above)
Comment (thirtiseven) [reply integration_tests/src/main/python/data_gen.py]: Thanks all! I also feel that the `hash_object` may not work as excepted at some edge cases and it is not easy to debug, but I didn't figured out a better solution then. 

Now I followed @revans2 's advice above to use `__repr__` to check the equality instead. 

My only concerns about this solution is it requires developers to implement `__repr__` of their custom datagens properly, so that it contains all the information that differs form others, otherwise there may be a collision that two different datagens are treated as the same by the cache.

**Comment by thirtiseven** [comment]
> > * maxsize = None: CacheInfo(hits=2820, misses=1023, maxsize=None, currsize=737)
> > * maxsize = 512: CacheInfo(hits=2788, misses=968, maxsize=512, currsize=512)
> > * maxsize = 256: CacheInfo(hits=3392, misses=1032, maxsize=256, currsize=256)
> > * maxsize = 128: CacheInfo(hits=4781, misses=945, maxsize=128, currsize=128)
> 
> Why is the number of hits decreasing when max_size is increasing?

I think it's because I didn't clear the cache after the test, somehow the datagens in lru_cache still seem to exist even all tests are end... I will update maxsize results later.

**Comment by thirtiseven** [comment]
build

**Comment by thirtiseven** [comment]
build

File: integration_tests/src/main/python/data_gen.py [hunk 4]
```diff
@@ -144,9 +154,9 @@ def __init__(self, child_gen, func, data_type=None, nullable=True):
         super().__init__(data_type, nullable=nullable)
         self._child_gen = child_gen
         self._func = func
-
+    
     def __repr__(self):
-        return super().__repr__() + '(' + str(self._child_gen) + ')'
+        return super().__repr__() + '(' + self._child_gen.__repr__() + ')'
```
[the remark this answers is not in the document]
Comment (thirtiseven) [reply integration_tests/src/main/python/data_gen.py:162]: No, but they need to override `__repr__`. `__repr__` will be part of the test case name, so if `specialcases` are placed there, the test case name will be very long and unreadable, so I used `cache_repr` to store them.

**Comment by thirtiseven** [comment]
build

**Comment by thirtiseven** [comment]
build

File: integration_tests/src/main/python/data_gen.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (thirtiseven) [reply integration_tests/src/main/python/data_gen.py]: Yes, It was a mistake, thank you! They need to override it because the nested types may also have their special cases. I removed the `cache_repr` now and hashed the `specialcases` to shorten the `__repr__`.

File: integration_tests/src/main/python/data_gen.py (hunk 2 above)
Comment (thirtiseven) [reply integration_tests/src/main/python/data_gen.py]: I did some tests, shows that the memory address (which python check for equality) of sparksession will reflect the current runtime configs. And since there is only one sparksession at the same time, we can use the memory address to cache spark with the same configs. I added a comment in `gen_df`.

File: integration_tests/src/main/python/data_gen.py [hunk 5]
```diff
@@ -681,6 +730,19 @@ def skip_if_not_utc():
     if (not is_tz_utc()):
         skip_unless_precommit_tests('The java system time zone is not set to UTC')
 
+# SparkSession with the same runtime configs will have same memory address and vice versa,
```
[the remark this answers is not in the document]
Comment (thirtiseven) [reply integration_tests/src/main/python/data_gen.py:733]: I recorded the max memory usage of a single test of all current IT, it's 7282688 bytes. so it may cost additional $128*7282688$ bytes $= 932$ MB memory if every slots with huge data sets. Not very large to cause an OOM, is it?

File: integration_tests/src/main/python/data_gen.py [hunk 6]
```diff
@@ -37,15 +39,23 @@ class DataGen:
...
+        # repr of DataGens and their children will be used to generate the cache key
+        # make sure it is unique for different DataGens
+        notnull = '(not_null)' if not self.nullable else ''
+        datatype = str(self.data_type)
+        specialcases = ''
+        for (weight, case) in self._special_cases:
+            if (callable(case)):
+                case = case.__code__.co_code
+            specialcases += str(case) + ' ' + str(weight) + ', '
+        specialcases = hashlib.blake2b(specialcases.encode('utf-8'), digest_size=8).hexdigest()
```
Comment (revans2) [inline integration_tests/src/main/python/data_gen.py:51]: This shows up in the name of the tests in many cases. I don't think I like having the special cases with a hash in it. I would much rather have a `_cache_repr` with all of the details that is separate from the normal __repr__, and then us the cache version for equality.

File: integration_tests/src/main/python/data_gen.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (pxLi) [reply integration_tests/src/main/python/data_gen.py:733]: sounds good to me,  can you also help add this numbers in comment for reference? so if we met any related issue people could try poke the cache size to quick workaround, thanks

[the remark this answers is not in the document]
Comment (thirtiseven) [reply integration_tests/src/main/python/data_gen.py:733]: Ok, I added it in description.

[the remark this answers is not in the document]
Comment (pxLi) [reply integration_tests/src/main/python/data_gen.py:733]: I mean add those in this change, maybe as a **NOTE** comment

**Comment by thirtiseven** [comment]
build

File: integration_tests/src/main/python/data_gen.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (thirtiseven) [reply integration_tests/src/main/python/data_gen.py:759]: Ok, added a comment here

**Comment by thirtiseven** [comment]
build

File: integration_tests/src/main/python/data_gen.py (hunk 2 above)
Comment (gerashegalov) [reply integration_tests/src/main/python/data_gen.py]: When spark.conf is mutated its address stays the same as well. But the createDataFrame result may differ. 

Here is a notebook with a contrived example demonstrating the issue https://github.com/gerashegalov/rapids-shell/blob/master/src/jupyter/create-dataframe-conf.ipynb

It would be safer to add spark._jconf.getAllConfs().toString() as an input to the hash key. Or have some logic checking that rejecting such confs at gen_df time.

Comment (thirtiseven) [reply integration_tests/src/main/python/data_gen.py]: I did more tests, you are right, thank you! Different configs do not affect the memory address, I was misled by the results of the parallel tests. I added `spark._jconf.getAllConfs().toString()` as cache key now. The results show that the cache hits decreased after this change, and I'm curious why all the IT passed when not considering this, will continue to find out why...

**Review by revans2 — DISMISSED** [review]

**Review by gerashegalov — APPROVED** [review]
LGTM

File: integration_tests/src/main/python/data_gen.py (hunk 2 above)
Comment (gerashegalov) [reply integration_tests/src/main/python/data_gen.py]: The change is conservatively pessimistic including the diffs in the configs that will not affect createDataFrame results.  I premised the thread on the need to reason which configs might be problematic.

> The results show that the cache hits decreased

One of the reasons is that with the pessimistic approach gen_df_help result is not cached between GPU and CPU since at least spark.rapids.sql.enabled values are different. It should be safe to ignore `spark.rapids.sql.enabled`  and `spark.rapids.sql.test.*` https://github.com/NVIDIA/spark-rapids/blob/291a389cc85ae8dbc609c50559de951a38b131da/integration_tests/src/main/python/spark_session.py#L112-L133

> why all the IT passed when not considering this

possible reasons
- current config variations don't cause issues 
- maybe tests for such edge cases just don't exist or scenarios not really valid
- we often compare results after converting them to Python-land arrays, it may cause discrepancies. In the contrived example `spark.sql('SELECT ts_ntz FROM df1').collect() == spark.sql('SELECT ts_ltz FROM df2').collect()` is True regardless of the conf while in SQL it will depend on the current timeZone.

Comment (thirtiseven) [reply integration_tests/src/main/python/data_gen.py]: Thanks, learned a lot. 

`SparkSession` object has no attribute `_jconf` in 3.1.1 so CI failed. So I think maybe we can cache the data list instead of caching dataframe for now to keep fine speedup without causing problems about sparkconf. 

I updated the code, please have a look. We can file a follow up issue about caching dataframe properly (considering the sparkconf issue) if you think caching data is ok for now.

Result for `TEST_PARALLEL=1 ./run_pyspark_from_build.sh -k 'cmp'` after this change: `430 passed, 19545 deselected, 1 warning in 199.97s (0:03:19)`

**Comment by thirtiseven** [comment]
build

**Comment by thirtiseven** [comment]
Misclick for request review form pxLi, please ignore.

**Comment by thirtiseven** [comment]
build

**Comment by thirtiseven** [comment]
build

File: integration_tests/src/main/python/data_gen.py (hunk 2 above)
Comment (gerashegalov) [reply integration_tests/src/main/python/data_gen.py]: +1 for a follow-up issue. I am fine going back to df caching in the meantime if it provides significant savings. but this LGTM too

**Review by gerashegalov — APPROVED** [review]

File: integration_tests/src/main/python/data_gen.py (hunk 2 above)
Comment (thirtiseven) [reply integration_tests/src/main/python/data_gen.py]: Filed follow-up issue #8524

Comment (revans2) [reply integration_tests/src/main/python/data_gen.py]: I don't understand how the configs are going to impact the data gen.  The data is not generated by Spark. The caching happens before the data is even handed to spark to produce a data frame.  We don't really support time zones yet, but when we do I would assume that it would be one of the parameters to the data gen.
